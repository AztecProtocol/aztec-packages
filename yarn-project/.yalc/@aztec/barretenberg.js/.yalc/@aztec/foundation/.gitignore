dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs