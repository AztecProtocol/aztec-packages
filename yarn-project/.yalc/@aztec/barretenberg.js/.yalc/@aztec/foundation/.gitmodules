[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
