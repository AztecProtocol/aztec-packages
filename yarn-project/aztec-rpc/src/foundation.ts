import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
