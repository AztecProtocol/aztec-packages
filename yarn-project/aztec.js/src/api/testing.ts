export { AnvilTestWatcher } from '../test/anvil_test_watcher.js';
export { RollupCheatCodes } from '../test/rollup_cheat_codes.js';
export { AztecCheatCodes } from '../test/aztec_cheat_codes.js';
export { EthCheatCodes } from '@aztec/ethereum/eth-cheatcodes';
export { CheatCodes } from '../test/cheat_codes.js';
