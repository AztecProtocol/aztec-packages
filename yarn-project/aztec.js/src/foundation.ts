export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export class Timer {
  private start: number;

  constructor() {
    this.start = new Date().getTime();
  }

  public ms() {
    return new Date().getTime() - this.start;
  }

  public s() {
    return (new Date().getTime() - this.start) / 1000;
  }
}

export function* backoffGenerator() {
  const v = [1, 1, 1, 2, 4, 8, 16, 32, 64];
  let i = 0;
  while (true) {
    yield v[Math.min(i++, v.length - 1)];
  }
}

export async function retry<Result>(fn: () => Promise<Result>, name = 'Operation', backoff = backoffGenerator()) {
  while (true) {
    try {
      return await fn();
    } catch (err: any) {
      const s = backoff.next().value;
      if (s === undefined) {
        throw err;
      }
      console.log(`${name} failed. Will retry in ${s}s...`);
      console.log(err);
      await sleep(s * 1000);
      continue;
    }
  }
}

// Call `fn` repeatedly until it returns true or timeout.
// Both `interval` and `timeout` are seconds.
// Will never timeout if the value is 0.
export async function retryUntil<T>(fn: () => Promise<T | undefined>, name = '', timeout = 0, interval = 1) {
  const timer = new Timer();
  while (true) {
    const result = await fn();
    if (result) {
      return result;
    }

    await sleep(interval * 1000);

    if (timeout && timer.s() > timeout) {
      throw new Error(name ? `Timeout awaiting ${name}` : 'Timeout');
    }
  }
}
