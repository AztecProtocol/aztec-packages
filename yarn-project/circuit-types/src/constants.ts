export const INITIAL_L2_BLOCK_NUM = 1;
export const BLOB_SIZE_IN_BYTES = 31 * 4096;
