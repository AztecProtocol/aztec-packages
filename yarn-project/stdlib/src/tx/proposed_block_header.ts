import { sha256ToField } from '@aztec/foundation/crypto';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import type { ZodFor } from '@aztec/foundation/schemas';
import { BufferReader, bigintToUInt64BE, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import type { FieldsOf } from '@aztec/foundation/types';

import { inspect } from 'util';
import { z } from 'zod';

import { AztecAddress } from '../aztec-address/index.js';
import { GasFees } from '../gas/index.js';
import { schemas } from '../schemas/index.js';
import type { UInt64 } from '../types/shared.js';
import { ContentCommitment } from './content_commitment.js';

/** The proposed values of an L2 block. */
export class ProposedBlockHeader {
  constructor(
    /** Root of the archive tree before this block is added. */
    public lastArchiveRoot: Fr,
    /** Content commitment of the L2 block. */
    public contentCommitment: ContentCommitment,
    /** Slot number of the L2 block */
    public slotNumber: Fr,
    /** Timestamp of the L2 block. */
    public timestamp: UInt64,
    /** Recipient of block reward. */
    public coinbase: EthAddress,
    /** Address to receive fees. */
    public feeRecipient: AztecAddress,
    /** Global gas prices for this block. */
    public gasFees: GasFees,
    /** Total mana used in the block, computed by the root rollup circuit */
    public totalManaUsed: Fr,
  ) {}

  static get schema(): ZodFor<ProposedBlockHeader> {
    return z
      .object({
        lastArchiveRoot: schemas.Fr,
        contentCommitment: ContentCommitment.schema,
        slotNumber: schemas.Fr,
        timestamp: schemas.BigInt,
        coinbase: schemas.EthAddress,
        feeRecipient: schemas.AztecAddress,
        gasFees: GasFees.schema,
        totalManaUsed: schemas.Fr,
      })
      .transform(ProposedBlockHeader.from);
  }

  static getFields(fields: FieldsOf<ProposedBlockHeader>) {
    return [
      fields.lastArchiveRoot,
      fields.contentCommitment,
      fields.slotNumber,
      fields.timestamp,
      fields.coinbase,
      fields.feeRecipient,
      fields.gasFees,
      fields.totalManaUsed,
    ] as const;
  }

  static from(fields: FieldsOf<ProposedBlockHeader>) {
    return new ProposedBlockHeader(...ProposedBlockHeader.getFields(fields));
  }

  static fromBuffer(buffer: Buffer | BufferReader): ProposedBlockHeader {
    const reader = BufferReader.asReader(buffer);

    return new ProposedBlockHeader(
      reader.readObject(Fr),
      reader.readObject(ContentCommitment),
      Fr.fromBuffer(reader),
      reader.readUInt64(),
      reader.readObject(EthAddress),
      reader.readObject(AztecAddress),
      reader.readObject(GasFees),
      reader.readObject(Fr),
    );
  }

  toBuffer() {
    // Note: The order here must match the order in the HeaderLib solidity library.
    return serializeToBuffer([
      this.lastArchiveRoot,
      this.contentCommitment,
      this.slotNumber,
      bigintToUInt64BE(this.timestamp),
      this.coinbase,
      this.feeRecipient,
      this.gasFees,
      this.totalManaUsed,
    ]);
  }

  hash(): Fr {
    return sha256ToField([this.toBuffer()]);
  }

  static empty(fields: Partial<FieldsOf<ProposedBlockHeader>> = {}): ProposedBlockHeader {
    return ProposedBlockHeader.from({
      lastArchiveRoot: Fr.ZERO,
      contentCommitment: ContentCommitment.empty(),
      slotNumber: Fr.ZERO,
      timestamp: 0n,
      coinbase: EthAddress.ZERO,
      feeRecipient: AztecAddress.ZERO,
      gasFees: GasFees.empty(),
      totalManaUsed: Fr.ZERO,
      ...fields,
    });
  }

  isEmpty(): boolean {
    return (
      this.lastArchiveRoot.isZero() &&
      this.contentCommitment.isEmpty() &&
      this.slotNumber.isZero() &&
      this.timestamp === 0n &&
      this.coinbase.isZero() &&
      this.feeRecipient.isZero() &&
      this.gasFees.isEmpty() &&
      this.totalManaUsed.isZero()
    );
  }

  /**
   * Serializes this instance into a string.
   * @returns Encoded string.
   */
  public toString() {
    return bufferToHex(this.toBuffer());
  }

  static fromString(str: string): ProposedBlockHeader {
    return ProposedBlockHeader.fromBuffer(hexToBuffer(str));
  }

  toInspect() {
    return {
      lastArchive: this.lastArchiveRoot.toString(),
      contentCommitment: this.contentCommitment.toInspect(),
      slotNumber: this.slotNumber.toBigInt(),
      timestamp: this.timestamp,
      coinbase: this.coinbase.toString(),
      feeRecipient: this.feeRecipient.toString(),
      gasFees: this.gasFees.toInspect(),
      totalManaUsed: this.totalManaUsed.toBigInt(),
    };
  }

  [inspect.custom]() {
    return `Header {
  lastArchiveRoot: ${this.lastArchiveRoot.toString()},
  contentCommitment: ${inspect(this.contentCommitment)},
  slotNumber: ${this.slotNumber.toBigInt()},
  timestamp: ${this.timestamp},
  coinbase: ${this.coinbase.toString()},
  feeRecipient: ${this.feeRecipient.toString()},
  gasFees: ${this.gasFees.toInspect()},
  totalManaUsed: ${this.totalManaUsed.toBigInt()},
}`;
  }
}
