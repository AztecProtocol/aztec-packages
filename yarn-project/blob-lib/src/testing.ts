import { BLOBS_PER_BLOCK } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';

import { Blob } from './blob.js';
import { BlobPublicInputs, BlockBlobPublicInputs } from './blob_public_inputs.js';
import { TX_START_PREFIX, TX_START_PREFIX_BYTES_LENGTH } from './encoding.js';
import { Poseidon2Sponge, SpongeBlob } from './sponge_blob.js';

/**
 * Makes arbitrary poseidon sponge for blob inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the sponge.
 * @returns A sponge blob instance.
 */
export function makeSpongeBlob(seed = 1): SpongeBlob {
  return new SpongeBlob(
    new Poseidon2Sponge(
      makeTuple(3, i => new Fr(i)),
      makeTuple(4, i => new Fr(i)),
      1,
      false,
    ),
    seed,
    seed + 1,
  );
}

/**
 * Makes arbitrary blob public inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the blob inputs.
 * @returns A blob public inputs instance.
 */
export function makeBlobPublicInputs(seed = 1): BlobPublicInputs {
  return new BlobPublicInputs(
    new Fr(seed),
    BigInt(seed + 1),
    makeTuple(2, i => new Fr(i)),
  );
}

/**
 * Makes arbitrary block blob public inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the blob inputs.
 * @returns A block blob public inputs instance.
 */
export function makeBlockBlobPublicInputs(seed = 1): BlockBlobPublicInputs {
  return new BlockBlobPublicInputs(makeTuple(BLOBS_PER_BLOCK, () => makeBlobPublicInputs(seed)));
}

// TODO: copied form stdlib tx effect
function encodeFirstField(length: number): Fr {
  const lengthBuf = Buffer.alloc(2);
  lengthBuf.writeUInt16BE(length, 0);
  return new Fr(
    Buffer.concat([
      toBufferBE(TX_START_PREFIX, TX_START_PREFIX_BYTES_LENGTH),
      Buffer.alloc(1),
      lengthBuf,
      Buffer.alloc(1),
      Buffer.from([1]),
      Buffer.alloc(1),
      Buffer.alloc(1),
    ]),
  );
}

/**
 * Make an encoded blob with the given length
 *
 * This will deserialise correctly in the archiver
 * @param length
 * @returns
 */
export function makeEncodedBlob(length: number): Promise<Blob> {
  return Blob.fromFields([encodeFirstField(length + 1), ...Array.from({ length: length }, () => Fr.random())]);
}

/**
 * Make an unencoded blob with the given length
 *
 * This will fail deserialisation in the archiver
 * @param length
 * @returns
 */
export function makeUnencodedBlob(length: number): Promise<Blob> {
  return Blob.fromFields([...Array.from({ length: length }, () => Fr.random())]);
}

export function makeEncodedBlobFields(fields: Fr[]): Promise<Blob> {
  return Blob.fromFields([encodeFirstField(fields.length + 1), ...fields]);
}
