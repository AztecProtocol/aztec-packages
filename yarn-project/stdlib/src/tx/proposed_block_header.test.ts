import { PROPOSED_BLOCK_HEADER_LENGTH_BYTES } from '@aztec/constants';
import { setupCustomSnapshotSerializers } from '@aztec/foundation/testing';
import { updateInlineTestData } from '@aztec/foundation/testing/files';

import { makeHeader } from '../tests/factories.js';
import { ProposedBlockHeader } from './proposed_block_header.js';

describe('ProposedBlockHeader', () => {
  let header: ProposedBlockHeader;

  beforeAll(() => {
    const seed = 9870243;
    setupCustomSnapshotSerializers(expect);
    header = makeHeader(seed, undefined).toPropose();
  });

  it('serializes to buffer and deserializes it back', () => {
    const buffer = header.toBuffer();
    expect(buffer.length).toBe(PROPOSED_BLOCK_HEADER_LENGTH_BYTES);
    const res = ProposedBlockHeader.fromBuffer(buffer);
    expect(res).toEqual(header);
  });

  it('computes hash', () => {
    const hash = header.hash();
    expect(hash).toMatchSnapshot();
  });

  it('computes empty hash', () => {
    const header = ProposedBlockHeader.empty();
    const hash = header.hash();
    expect(hash).toMatchSnapshot();

    // Run with AZTEC_GENERATE_TEST_DATA=1 to update noir test data
    updateInlineTestData(
      'noir-projects/noir-protocol-circuits/crates/types/src/proposed_block_header.nr',
      'test_data_empty_hash',
      hash.toString(),
    );
  });
});
