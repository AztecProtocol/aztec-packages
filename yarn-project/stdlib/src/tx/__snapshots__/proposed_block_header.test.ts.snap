// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposedBlockHeader computes empty hash 1`] = `Fr<0x002e384af86a480f952aa16443fd29646a9063865e62d7c403fc7ed697bb7712>`;

exports[`ProposedBlockHeader computes hash 1`] = `Fr<0x00f99d951c77fcee0f9806913dfbcc004ba7db1181ff74c5727b22b341b016ea>`;
